### PR TITLE
Speed up condition normalization (DNF conversion)

### PR DESCRIPTION
**Pull request by danielhlarkin**

### CODE REVIEW

File: arangod/Aql/AstNode.cpp [hunk 1]
```diff
@@ -2620,6 +2622,18 @@ void AstNode::removeMembersInOtherAndNode(AstNode const* other) {
   }
 }
 
+void AstNode::markFinalized(AstNode const* subtreeRoot) {
```
Comment (jsteemann) [inline arangod/Aql/AstNode.cpp]: It is really-counter intuitive that `markFinalized` is getting a const AstNode as its input.
Can we pass a non-const input to it?

**Comment by jsteemann** [comment]
AFAIK the "replaceGeoCondition" function in OptimizerRules.cpp will modify nodes in place, but it does not temporarily unlock them if I am not wrong.
Do we need to add some unlocking there too?

File: arangod/Aql/AstNode.cpp (hunk 1 above)
Comment (danielhlarkin) [reply arangod/Aql/AstNode.cpp]: Will fix.

**Comment by danielhlarkin** [comment]
@jsteemann So the `replaceGeoCondition` method seems to primarily do modifications via `traverseAndModify`, which does the unlocking internally when necessary. Are you referring to the end where it modifies `replaceInfo.expressionParent`? I'll added unlocking there. Let me know if you mean somewhere else as well.

**Comment by jsteemann** [comment]
@danielhlarkin : yes, that was the place I was referring to. Maybe I overlooked the unlocking there. Nevermind.

File: arangod/Transaction/Methods.cpp [hunk 2]
```diff
@@ -314,8 +314,21 @@ bool transaction::Methods::sortOrs(
...
+        // must edit nodes in place; TODO change so we can replace with copy
+        auto n1 = root->getMember(previousIn)->getMember(0);
+        n1->flags = (n1->flags &
+                     ~arangodb::aql::AstNodeFlagType::FLAG_FINALIZED);
+
+        TRI_DEFER(FINALIZE_SUBTREE(n1));
+        n1->changeMember(1, mergedIn);
+
+        auto n2 = root->getMember(i)->getMember(0);
+        n2->flags = (n2->flags &
```
[the remark this answers is not in the document]
Comment (danielhlarkin) [reply arangod/Transaction/Methods.cpp]: We could, though I think for when we unset the finalized flag, we are essentially breaking the API anyway and thus should probably not codify this usage in the API. Hence, wrap it in an ugly macro that we will work on replacing (soon?) in a separate PR where we get rid of the unlocking. Or better yet, where we make the nodes properly immutable and get rid of the flag altogether.

File: arangod/Transaction/Methods.cpp [hunk 3]
```diff
@@ -314,8 +314,21 @@ bool transaction::Methods::sortOrs(
             parts[previousIn].valueNode, p.valueNode);
         parts[previousIn].valueNode = mergedIn;
         parts[i].valueNode = emptyArray;
-        root->getMember(previousIn)->getMember(0)->changeMember(1, mergedIn);
-        root->getMember(i)->getMember(0)->changeMember(1, emptyArray);
+
+        // must edit nodes in place; TODO change so we can replace with copy
+        auto n1 = root->getMember(previousIn)->getMember(0);
```
[the remark this answers is not in the document]
Comment (danielhlarkin) [reply arangod/Transaction/Methods.cpp]: Ah, I forgot to go replace that pattern with `TEMPORARILY_UNLOCK_NODE`. Will fix.

It's not great, but essentially, when we want to change a node, we must also replace all its ancestors. The current API for the method does not allow us to change/return the root that is passed in. I did not want to make major changes in this PR, so I added the guarded unlocking mechanism to skirt around the issue for now. It does not cause problems for the current implementation (at least not in anything in our test suite), but should definitely be properly fixed soon in another PR.

**Pull request merged**